sympy/physics/quantum/tensorproduct.py

```diff
@@ -246,9 +246,11 @@ def _eval_expand_tensorproduct(self, **hints):
             if isinstance(args[i], Add):
                 for aa in args[i].args:
                     tp = TensorProduct(*args[:i] + (aa,) + args[i + 1:])
-                    if isinstance(tp, TensorProduct):
-                        tp = tp._eval_expand_tensorproduct()
-                    add_args.append(tp)
+                    c_part, nc_part = tp.args_cnc()
+                    #Check for TensorProduct object: is the one object in nc_part, if any:
+                    if len(nc_part)==1 and isinstance(nc_part[0], TensorProduct):
```

Comment:
What happens if there are other thing in the nc_part?

Reply:
That is a very valid remark!
Yes, strictly speaking, we would have to do the recursion in full: call _eval_expand_tensorproduct on all TensorProduct objects found in nc_args and call .expand(**hints)  on the product of this and the other objects in nc_part!

I wanted to avoid the danger of circcular recursion. Looking into the constructor of TensorProduct, we see that matrices are handled in their own way by matrix_tensor_product, while for TensorProducts it would require a special subclass  that constructs products of nc-factors and TensorProducts to make situation 3 below occur. 

So the situation where nc_parts has non-TensorProduct objects or multiple objects seems so rare that I decided not to solve it completely but to gracefully handle it:

1) If the 'if-then' branch is not taken, the contents of tp will be reconstructed in the subsequent line 'add_args.append' by the 'Mul(*c_part)*Mul(*nc_part)'. This ensures that the  output is mathematically correct. 
2) The same is true if nc_part has exactly one object, and this object is not TensorProuduct: The result is correct, but this one object is not expanded
3) If nc_part has two or more objects, I again handle this as a don't-know-how-to-deal-with situation and take the defensive way out: the if-then is not taken, and the subsequent line ensures that the output is mathematically correct, though the expansion may be incomplete, so the damage is limited.

So I believe the simplification I've taken is efficient in the sense that it clearly extends the capabilities using very controlled and testable means. 

Of course the 'full recursion' is possible, however I'm unsure how to verify it.

Reply:
Regarding test cases: yes, of course.   I'll put them into test_tensorproduct.py, into test_tensor_product_expand()